src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.ICSharpProjectSite.cs

```diff
@@ -124,9 +123,14 @@ public int GetValidStartupClasses(IntPtr[] classNames, ref int count)
             // If classNames is NULL, then we need to populate the number of valid startup
             // classes only
             var project = Workspace.CurrentSolution.GetProject(VisualStudioProject.Id);
-            var compilation = project.GetCompilationAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None);
+            if (project == null)
+            {
+                count = 0;
+                return VSConstants.S_OK;
+            }
 
-            var entryPoints = EntryPointFinder.FindEntryPoints(compilation.Assembly.GlobalNamespace);
+            var compilation = project.GetRequiredCompilationAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None);
```

Comment:
are we in contact with the team that owns this API?  this is a really good example of something that should be async as this call is enormously expensive.

Reply:
@jasonmalinowski I believe this is the old csproj code correct? Presumably if the project system ever moves to be entirely CPS-based we will be able to fix this.

Reply:
Indeed, this is for the legacy project systems. That said, it's possible there's an easy way to have the project system call us here on a background thread, or make it async even in the legacy system.

Reply:
Is there a plan written down somewhere for when we are going to audit and remove https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/TaskExtensions.cs in general @sharwell? Presumably we want to delete this code and either use JoinableTaskFactory or some other api in production code?

@jasonmalinowski It appears that we are called through COM from C++ in [this case](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=/src/vsproject/cool/coolbld/CCSharpBuildCompiler.cpp&version=GBmain&line=1352&lineEnd=1375&lineStartColumn=1&lineEndColumn=2&lineStyle=plain&_a=contents). I am unfamiliar with making async COM calls from native to managed. Can you point me to a place where we've done this before?

Reply:
i would not want to use JTF here.  i'm basically saying that the entire end-to-end experience (including for the customer) Would be that this was async :)

Reply:
> I am unfamiliar with making async COM calls from native to managed. Can you point me to a place where we've done this before?

There is an IVsTask which is a native/COM version of a task-like API which you can interop with. I'm not sure we have any examples in Roslyn. What might be easier though is if the API here can be called on a background thread, then don't make any changes on the Roslyn side and instead just update the project system to jump to a background thread and then call it. The challenge though is there's a number of layers in the middle there that you'd have to thread through though.

Reply:
Alright, I'll investigate this. Will submit this in a follow-up PR if possible

Reply:
@jmarolf Yeah, don't try to do this now in 17.2. That won't end well. :smile: